metabase/metabase
# Use sessions in mongo aggregation pipelines

Closes #32372

### Description

Prior to this PR mongo driver did not kill the queries when connection was closed (ie. browser windows was closed). This PR takes advantage mongo's sessions to accomplish that. 

_Use of sessions._ We were not using sessions in an explicit way in the mongo driver prior to this PR. From what I've found, use explicit sessions provides the only mechanism capable of killing a running operation. For more context see [this section](https://metaboat.slack.com/docs/T078VCLCR/F062AFF81K9?focus_section_id=temp:C:ITf577344081b0d44a48022f5070) of my notes. Specifically links to [the issue](https://jira.mongodb.org/browse/JAVA-1109) on mongodb's JIRA and this stackoverflow [question](https://stackoverflow.com/questions/31079479/cancel-a-mongodb-query-from-java-api). Also [the sessions specification](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#explicit-vs-implicit-sessions) describes killing of long running operations by the means of sessions.

_Performance impact_. I've measured the average execution time of 100 basic queries before and after the change, finding no significant difference.

_Restrictions on mongo topologies (deployment setups)_. Sessions were introduced in mongodb version 3.6. We currently support versions 4.2+. [Mongodb reference](https://www.mongodb.com/docs/manual/reference/server-sessions/#overview) states that:
> Starting in 3.6, MongoDB drivers associate all operations with a server session...

but in contrary to that, previous section of that reference states following:
> Server sessions are available for replica sets and sharded clusters only.

I've asked for the correct interpretation of the cited sentence on relevant fora with no answer to this point. However, testing (and development) on our docker images shown that no clustering or replication setup was necessary for the sessions to work properly. Hence, I believe we are safe to go further with this change. Furthermore, no relevant restrictions on mongo topologies were found.

_Useful links_:
- [Mongodb sessions spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst).
- [Mongodb server sessions reference](https://www.mongodb.com/docs/manual/reference/server-sessions/).
- [Driver's javadoc 3.12](https://mongodb.github.io/mongo-java-driver/3.12/javadoc/).

### How to verify

See the new test, or see mongo logs after closing browser window with with in progress query.

### Checklist
